File: dungeons-1.21.11/src/main/java/fishmod/features/dungeon/f7/PillarExplode.java
package fishmod.features.dungeon.f7;

import config.practical.hud.HUDComponent;
import fishmod.utils.Constants;
import fishmod.utils.Location;
import fishmod.utils.Scheduler;
import fishmod.utils.config.values.Floor7;
import fishmod.utils.dungeon.Phase;
import fishmod.utils.events.Events;
import fishmod.utils.rendering.RenderUtils;
import net.minecraft.client.gui.DrawContext;
import net.minecraft.sound.SoundEvents;
import net.minecraft.text.Text;

/** "Storm crushed!" notification + optional pillar-explode timer. Ported from blade-addons. */
public class PillarExplode {

    private static final int TOTAL_TICKS = 20;
    private static int tick = 0;

    public static void init() {
        Events.ON_GAME_MESSAGE.register(text -> {
            if (!Floor7.notifyStormCrush && !Floor7.timePillarExplosion) return false;
            if (!Location.inDungeon() || !Phase.inP2()) return false;
            String string = text.getString();
            if (string == null) return false;
            if (string.equals("[BOSS] Storm: Oof") || string.equals("[BOSS] Storm: Ouch, that hurt!")) {
                tick = TOTAL_TICKS;
                Scheduler.scheduleSound(SoundEvents.BLOCK_NOTE_BLOCK_PLING.value(), 1, 1);
            }
            return false;
        });
        Events.ON_SERVER_TICK.register(() -> { tick = Math.max(tick - 1, 0); return false; });
    }

    public static boolean displayTimer() { return Floor7.timePillarExplosion && tick > 0; }

    public static void renderTimer(HUDComponent component, DrawContext context) {
        RenderUtils.drawTimer(component, context, tick, tick < 6 ? Constants.GREEN : Constants.RED);
    }

    public static boolean display() { return Floor7.notifyStormCrush && tick > 0; }

    public static void render(HUDComponent component, DrawContext context) {
        RenderUtils.drawCenteredText(context, component, Text.literal("§6||| §bStorm crushed! §6|||"));
    }
}

File: dungeons-26.1.2/src/main/kotlin/fishmod/features/dungeon/f7/PillarExplode.kt
package fishmod.features.dungeon.f7

import config.practical.hud.HUDComponent
import fishmod.utils.Constants
import fishmod.utils.Location
import fishmod.utils.Scheduler
import fishmod.utils.config.values.Floor7
import fishmod.utils.dungeon.Phase
import fishmod.utils.events.Events
import fishmod.utils.rendering.RenderUtils
import net.minecraft.client.gui.GuiGraphicsExtractor
import net.minecraft.network.chat.Component
import net.minecraft.sounds.SoundEvents

/** "Storm crushed!" notification + optional pillar-explode timer. Ported from blade-addons. */
object PillarExplode {

    private const val TOTAL_TICKS = 20
    private var tick = 0

    @JvmStatic
    fun init() {
        Events.ON_GAME_MESSAGE.register { text ->
            if (!Floor7.notifyStormCrush && !Floor7.timePillarExplosion) return@register false
            if (!Location.inDungeon() || !Phase.inP2()) return@register false
            val string = text.string
            if (string == null) return@register false
            if (string == "[BOSS] Storm: Oof" || string == "[BOSS] Storm: Ouch, that hurt!") {
                tick = TOTAL_TICKS
                Scheduler.scheduleSound(SoundEvents.NOTE_BLOCK_PLING.value(), 1f, 1f)
            }
            false
        }
        Events.ON_SERVER_TICK.register {
            tick = maxOf(tick - 1, 0)
            false
        }
    }

    @JvmStatic
    fun displayTimer(): Boolean = Floor7.timePillarExplosion && tick > 0

    @JvmStatic
    fun renderTimer(component: HUDComponent, context: GuiGraphicsExtractor) {
        RenderUtils.drawTimer(component, context, tick, if (tick < 6) Constants.GREEN else Constants.RED)
    }

    @JvmStatic
    fun display(): Boolean = Floor7.notifyStormCrush && tick > 0

    @JvmStatic
    fun render(component: HUDComponent, context: GuiGraphicsExtractor) {
        RenderUtils.drawCenteredText(context, component, Component.literal("§6||| §bStorm crushed! §6|||"))
    }
}
